src/utils.rs

```diff
@@ -0,0 +1,280 @@
+use crate::error::SlateDBError;
+use crate::error::SlateDBError::{BackgroundTaskFailed, BackgroundTaskShutdown};
+use std::future::Future;
+
+pub(crate) struct WriteOnceRegister<T: Clone> {
+    rx: tokio::sync::watch::Receiver<Option<T>>,
+    tx: tokio::sync::watch::Sender<Option<T>>,
+}
+
+pub(crate) struct WriteOnceRegisterReader<T: Clone> {
+    rx: tokio::sync::watch::Receiver<Option<T>>,
+}
+
+impl<T: Clone> WriteOnceRegister<T> {
+    pub(crate) fn new() -> Self {
+        let (tx, rx) = tokio::sync::watch::channel(None);
+        Self { rx, tx }
+    }
+
+    pub(crate) fn write(&self, val: T) {
+        self.tx.send_if_modified(|v| {
+            if v.is_some() {
+                return false;
+            }
+            v.replace(val);
+            true
+        });
+    }
+
+    pub(crate) fn reader(&self) -> WriteOnceRegisterReader<T> {
+        WriteOnceRegisterReader {
+            rx: self.rx.clone(),
+        }
+    }
+}
+
+impl<T: Clone> WriteOnceRegisterReader<T> {
+    pub(crate) fn read(&self) -> Option<T> {
+        self.rx.borrow().clone()
+    }
+
+    pub(crate) async fn await_value(&mut self) -> T {
+        loop {
+            if let Some(maybe_value) = self.rx.borrow().as_ref() {
+                return maybe_value.clone();
+            }
+            self.rx.changed().await.expect("watch channel closed")
+        }
+    }
+}
+
+pub(crate) fn map_stopped_task_result<T>(
+    result: &Result<T, SlateDBError>,
+) -> Result<(), SlateDBError> {
+    match result {
+        Ok(_) => Err(BackgroundTaskShutdown),
+        Err(err) => Err(err.clone()),
+    }
+}
+
+pub(crate) fn spawn_bg_task<F, T, C>(
+    handle: &tokio::runtime::Handle,
+    cleanup_fn: C,
+    future: F,
+) -> tokio::task::JoinHandle<Result<T, SlateDBError>>
+where
+    F: Future<Output = Result<T, SlateDBError>> + Send + 'static,
+    T: Send + 'static,
+    C: FnOnce(&SlateDBError) + Send + 'static,
+{
+    let inner_handle = handle.clone();
+    handle.spawn(async move {
+        let jh = inner_handle.spawn(future);
```

Comment:
why not directly use `match future.await` here?

Reply:
I wouldn't be able to observe panics in that case. By spawning a task and joining it, I get an error back when the spawned task panics.